### ISSUE TITLE
SPSA has an assert that is too loose

### ISSUE THREAD

**Issue by goodfeli** [cleverhans-lab/cleverhans#801]
Grep for "TODO: this assert looks totally wrong."

**Comment by GuanxiongLiu** [comment]
Is it the same assertion error that I meet?
```
from cleverhans.attacks import SPSA
SPSA_params = {
    'epsilon': 0.6,
    'num_steps': 40
}
################
SPSA_adv = SPSA(adv, sess=sess)
SPSA_adv_gen = SPSA_adv.generate(x, **SPSA_params)
adv_SPSA = adv.get_logits(SPSA_adv_gen)
adv_SPSA_acc = model_eval(sess, x, y, adv_SPSA, X_test, Y_test, args=eval_params)
print('Adv training model accuracy on SPSA test examples: %0.4f' % adv_SPSA_acc)
```

/usr/local/lib/python3.6/dist-packages/cleverhans/attacks_tf.py in _compute_gradients(self, loss_fn, x, unused_optim_state)
   1650         """Compute gradient estimates using SPSA."""
   1651         # Assumes `x` is a list, containing a [1, H, W, C] image
-> 1652         assert len(x) == 1 and x[0].get_shape().as_list()[0] == 1
   1653         x = x[0]
   1654         x_shape = x.get_shape().as_list()

AssertionError:

**Comment by goodfeli** [comment]
That's a different assert

**Comment by jonasguan** [comment]
Closing as we update the library to v4 and drop support for version < 4.
